component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -199,4 +212,119 @@ private List<String> populateMandatoryClaims() {
+            }
+            if (!(isNotEmpty(requestSigningAlgorithm) &&
+                    requestSigningAlgorithm.equals(registeredSigningAlgorithm))) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Registered algorithm does not match with the token signed algorithm");
+                }
+                return false;
+            } else {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Obtain the request signing algorithms registered for the application.
+     *
+     * @param clientId   Client ID of the application.
+     * @return Registered signing algorithm for the application.
+     * @throws OAuthClientAuthnException
+     */
+    private String getRegisteredSigningAlgorithm(String clientId) throws OAuthClientAuthnException {
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (!(isNotEmpty(serviceProvider.getCertificateContent()))) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (Constants.TOKEN_ENDPOINT_AUTH_SIGNING_ALG.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+        return Constants.NOT_APPLICABLE;
+    }
+
+    /**
+     * Obtain the request signing algorithm from the signed JWT.
+     *
+     * @param signedObject   Signed JWT in the request.
+     * @return Request signing algorithm.
+     * @throws OAuthClientAuthnException
+     */
+    private String getRequestSigningAlgorithm(String signedObject) throws OAuthClientAuthnException {
+        try {
+            SignedJWT signedJWT = SignedJWT.parse(signedObject);
+            return signedJWT.getHeader().getAlgorithm().getName();
+        } catch (ParseException e) {
+            throw new OAuthClientAuthnException("Error occurred while parsing the signed assertion",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+    }
+
+    /**
+     * Check whether the application should support FAPI.
+     *
+     * @param clientId       Client ID of the application.
+     * @return Whether the application should support FAPI.
+     * @throws OAuthClientAuthnException
+     */
+    private String isFapiApplication(String clientId) throws OAuthClientAuthnException {
+
+        String isFapiApp = "IsFAPIApp";
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (isEmpty(serviceProvider.getCertificateContent())) {
```

Comment:
Why do we need to check whether the certificate content is empty?

Reply:
Fixed in c8efc7a5d86d1df217e9bdcab4167db2d8a20ab5